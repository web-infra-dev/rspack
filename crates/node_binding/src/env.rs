use std::env;
use std::ffi::OsStr;

pub fn set_env_if_unset<K, V>(k: K, v: V)
where
  K: AsRef<OsStr>,
  V: AsRef<OsStr>,
{
  match env::var(&k) {
    Ok(_) => (),
    Err(_) => {
      env::set_var(&k, v);
    }
  }
}
