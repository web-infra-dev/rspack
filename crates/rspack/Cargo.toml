[package]
edition    = "2021"
license    = "MIT"
name       = "rspack"
repository = "https://github.com/web-infra-dev/rspack"
version    = "0.1.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dev-dependencies]
criterion              = { version = "0.3.6", features = ["async_tokio", "async_futures"] }
insta                  = { workspace = true }
rspack_binding_options = { path = "../rspack_binding_options" }
rspack_testing         = { path = "../rspack_testing" }
serde                  = { workspace = true, features = ["derive"] }
serde_json             = { workspace = true }
testing_macros         = { workspace = true }
ustr                   = { workspace = true }
xshell                 = "0.2.2"

[target.'cfg(all(not(all(target_os = "linux", target_arch = "aarch64", target_env = "musl"))))'.dev-dependencies]
mimalloc-rust = { workspace = true }

[dependencies]
crossbeam                         = { workspace = true }
dashmap                           = { workspace = true }
rspack_core                       = { path = "../rspack_core" }
rspack_fs                         = { path = "../rspack_fs", features = ["async", "rspack-error"] }
rspack_ids                        = { path = "../rspack_ids" }
rspack_loader_sass                = { path = "../rspack_loader_sass" }
rspack_plugin_asset               = { path = "../rspack_plugin_asset" }
rspack_plugin_copy                = { path = "../rspack_plugin_copy" }
rspack_plugin_css                 = { path = "../rspack_plugin_css" }
rspack_plugin_devtool             = { path = "../rspack_plugin_devtool" }
rspack_plugin_externals           = { path = "../rspack_plugin_externals" }
rspack_plugin_html                = { path = "../rspack_plugin_html" }
rspack_plugin_javascript          = { path = "../rspack_plugin_javascript" }
rspack_plugin_json                = { path = "../rspack_plugin_json" }
rspack_plugin_remove_empty_chunks = { path = "../rspack_plugin_remove_empty_chunks" }
rspack_plugin_runtime             = { path = "../rspack_plugin_runtime" }
rspack_symbol                     = { path = "../rspack_symbol" }
# futures = "0.3.21"
anyhow = { workspace = true }
async-trait = { workspace = true }
once_cell = { workspace = true }
rspack_error = { path = "../rspack_error" }
rspack_tracing = { path = "../rspack_tracing" }
tokio = { workspace = true, features = [
  "rt",
  "rt-multi-thread",
  "macros",
  "test-util",
  "parking_lot",
] }
tracing = { workspace = true }
tracing-subscriber = { workspace = true, features = ["env-filter"] }
[[bench]]
harness = false
name    = "main"
