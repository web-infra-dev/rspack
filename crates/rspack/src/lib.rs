#![recursion_limit = "256"]
pub use rspack_core::Compiler;
use rspack_core::{CompilerOptions, Plugin, TargetPlatform};
use rspack_error::Result;
use rspack_plugin_asset::AssetConfig;
use rspack_plugin_devtool::DevtoolPluginOptions;

pub fn rspack(mut options: CompilerOptions, mut plugins: Vec<Box<dyn Plugin>>) -> Compiler {
  // Css plugin is registered via option normalization
  plugins.push(Box::new(rspack_plugin_asset::AssetPlugin::new(
    AssetConfig {
      parse_options: options.module.parser.as_ref().and_then(|x| x.asset.clone()),
    },
  )));
  plugins.push(Box::new(rspack_plugin_json::JsonPlugin {}));
  match &options.target.platform {
    TargetPlatform::Web => {
      plugins.push(Box::new(
        rspack_plugin_runtime::ArrayPushCallbackChunkFormatPlugin {},
      ));
      plugins.push(Box::new(rspack_plugin_runtime::RuntimePlugin {}));
      plugins.push(Box::new(rspack_plugin_runtime::CssModulesPlugin {}));
      plugins.push(Box::new(rspack_plugin_runtime::JsonpChunkLoadingPlugin {}));
    }
    TargetPlatform::Node(_) => {
      plugins.push(Box::new(
        rspack_plugin_runtime::CommonJsChunkFormatPlugin {},
      ));
      plugins.push(Box::new(rspack_plugin_runtime::RuntimePlugin {}));
      plugins.push(Box::new(
        rspack_plugin_runtime::CommonJsChunkLoadingPlugin {},
      ));
    }
    _ => {
      plugins.push(Box::new(rspack_plugin_runtime::RuntimePlugin {}));
    }
  };
  if options.dev_server.hot {
    plugins.push(Box::new(
      rspack_plugin_runtime::HotModuleReplacementPlugin {},
    ));
  }
  plugins.push(Box::new(
    rspack_plugin_runtime::BasicRuntimeRequirementPlugin {},
  ));
  if options.experiments.lazy_compilation {
    plugins.push(Box::new(rspack_plugin_runtime::LazyCompilationPlugin {}));
  }
  plugins.push(Box::new(rspack_plugin_externals::ExternalPlugin::default()));
  plugins.append(&mut options.plugins);
  plugins.push(Box::new(rspack_plugin_javascript::JsPlugin::new()));
  plugins.push(Box::new(rspack_plugin_devtool::DevtoolPlugin::new(
    DevtoolPluginOptions {
      inline: options.devtool.inline(),
      append: !options.devtool.hidden(),
      namespace: options.output.unique_name.clone(),
      columns: !options.devtool.cheap(),
      no_sources: options.devtool.no_sources(),
      public_path: None,
    },
  )));
  match options.module_ids {
    rspack_core::ModuleIds::Named => {
      plugins.push(Box::new(rspack_ids::NamedModuleIdsPlugin {}));
    }
    rspack_core::ModuleIds::Deterministic => {
      plugins.push(Box::new(rspack_ids::DeterministicModuleIdsPlugin {}));
    }
  }

  Compiler::new(options, plugins)
}

pub fn dev_server(options: CompilerOptions, plugins: Vec<Box<dyn Plugin>>) -> DevServer {
  DevServer {
    compiler: rspack(options, plugins),
  }
}

pub struct DevServer {
  compiler: Compiler,
}

impl DevServer {
  pub async fn serve(&mut self) -> Result<()> {
    self.compiler.build().await?;

    warp::fs::dir(self.compiler.options.context.join("dist"));
    let filter = warp::fs::dir(self.compiler.options.context.join("dist"));

    // let routes = warp::ws().map(|ws: warp::ws::Ws| {
    //   // And then our closure will be called when it completes...
    //   ws.on_upgrade(|websocket| {
    //     // Just echo all messages back...
    //     let (tx, rx) = websocket.split();
    //     rx.forward(tx).map(|result| {
    //       if let Err(e) = result {
    //         eprintln!("websocket error: {:?}", e);
    //       }
    //     })
    //   })
    // });

    warp::serve(filter).run(([127, 0, 0, 1], 3031)).await;
    Ok(())
  }
}
