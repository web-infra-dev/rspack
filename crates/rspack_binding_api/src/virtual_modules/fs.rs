use std::{
  fmt::Debug,
  sync::{Arc, RwLock},
};

use rspack_fs::{FileMetadata, FilePermissions, ReadableFileSystem, Result};
use rspack_paths::{Utf8Path, Utf8PathBuf};

use crate::virtual_modules::VirtualFileStore;

pub struct VirtualFileSystem {
  real_fs: Arc<dyn ReadableFileSystem>,
  virtual_file_store: Arc<RwLock<dyn VirtualFileStore>>,
}

impl VirtualFileSystem {
  pub fn new(
    real_fs: Arc<dyn ReadableFileSystem>,
    virtual_file_store: Arc<RwLock<dyn VirtualFileStore>>,
  ) -> Self {
    Self {
      real_fs,
      virtual_file_store,
    }
  }
}

impl Debug for VirtualFileSystem {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    f.debug_struct("VirtualFileSystem")
      .field("virtual_file_store", &"<VirtualFileStore>")
      .field("real_fs", &"<ReadableFileSystem>")
      .finish()
  }
}

#[async_trait::async_trait]
impl ReadableFileSystem for VirtualFileSystem {
  async fn read(&self, path: &Utf8Path) -> Result<Vec<u8>> {
    if let Ok(store) = self.virtual_file_store.read()
      && let Some(content) = store.get_file_content(path)
    {
      return Ok(content.clone());
    }

    self.real_fs.read(path).await
  }

  fn read_sync(&self, path: &Utf8Path) -> Result<Vec<u8>> {
    if let Ok(store) = self.virtual_file_store.read()
      && let Some(content) = store.get_file_content(path)
    {
      return Ok(content.clone());
    }

    self.real_fs.read_sync(path)
  }

  async fn metadata(&self, path: &Utf8Path) -> Result<FileMetadata> {
    if let Ok(store) = self.virtual_file_store.read()
      && let Some(metadata) = store.get_file_metadata(path)
    {
      return Ok(metadata.clone());
    }

    self.real_fs.metadata(path).await
  }

  fn metadata_sync(&self, path: &Utf8Path) -> Result<FileMetadata> {
    if let Ok(store) = self.virtual_file_store.read()
      && let Some(metadata) = store.get_file_metadata(path)
    {
      return Ok(metadata.clone());
    }

    self.real_fs.metadata_sync(path)
  }

  async fn symlink_metadata(&self, path: &Utf8Path) -> Result<FileMetadata> {
    if let Ok(store) = self.virtual_file_store.read()
      && let Some(metadata) = store.get_file_metadata(path)
    {
      return Ok(metadata.clone());
    }

    self.real_fs.symlink_metadata(path).await
  }

  async fn canonicalize(&self, path: &Utf8Path) -> Result<Utf8PathBuf> {
    if let Ok(store) = self.virtual_file_store.read()
      && store.contains(path)
    {
      return Ok(path.to_path_buf());
    }

    self.real_fs.canonicalize(path).await
  }

  async fn read_dir(&self, dir: &Utf8Path) -> Result<Vec<String>> {
    if let Some(mut vlist) = self
      .virtual_file_store
      .read()
      .ok()
      .and_then(|store| store.read_dir(dir))
    {
      let mut list = self.real_fs.read_dir(dir).await.unwrap_or_default();
      list.append(&mut vlist);
      Ok(list)
    } else {
      self.real_fs.read_dir(dir).await
    }
  }

  fn read_dir_sync(&self, dir: &Utf8Path) -> Result<Vec<String>> {
    if let Some(mut vlist) = self
      .virtual_file_store
      .read()
      .ok()
      .and_then(|store| store.read_dir(dir))
    {
      let mut list = self.real_fs.read_dir_sync(dir).unwrap_or_default();
      list.append(&mut vlist);
      Ok(list)
    } else {
      self.real_fs.read_dir_sync(dir)
    }
  }

  async fn permissions(&self, path: &Utf8Path) -> Result<Option<FilePermissions>> {
    if let Ok(store) = self.virtual_file_store.read()
      && store.contains(path)
    {
      return Ok(Some(FilePermissions::from_mode(0o700)));
    }

    self.real_fs.permissions(path).await
  }
}
