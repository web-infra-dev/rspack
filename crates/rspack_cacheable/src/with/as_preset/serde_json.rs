use rkyv::{
  rancor::Fallible,
  ser::Writer,
  string::{ArchivedString, StringResolver},
  with::{ArchiveWith, DeserializeWith, SerializeWith},
  Place,
};
use serde_json::Value;

use super::AsPreset;
use crate::{DeserializeError, SerializeError};

pub struct SerdeJsonResolver {
  inner: StringResolver,
  value: String,
}

impl ArchiveWith<Value> for AsPreset {
  type Archived = ArchivedString;
  type Resolver = SerdeJsonResolver;

  #[inline]
  fn resolve_with(_field: &Value, resolver: Self::Resolver, out: Place<Self::Archived>) {
    let SerdeJsonResolver { inner, value } = resolver;
    ArchivedString::resolve_from_str(&value, inner, out);
  }
}

impl<S> SerializeWith<Value, S> for AsPreset
where
  S: Fallible<Error = SerializeError> + Writer,
{
  #[inline]
  fn serialize_with(field: &Value, serializer: &mut S) -> Result<Self::Resolver, SerializeError> {
    let value = serde_json::to_string(field)
      .map_err(|_| SerializeError::MessageError("serialize serde_json value failed"))?;
    let inner = ArchivedString::serialize_from_str(&value, serializer)?;
    Ok(SerdeJsonResolver { value, inner })
  }
}

impl<D> DeserializeWith<ArchivedString, Value, D> for AsPreset
where
  D: Fallible<Error = DeserializeError>,
{
  #[inline]
  fn deserialize_with(field: &ArchivedString, _: &mut D) -> Result<Value, DeserializeError> {
    serde_json::from_str(field)
      .map_err(|_| DeserializeError::MessageError("deserialize serde_json value failed"))
  }
}
