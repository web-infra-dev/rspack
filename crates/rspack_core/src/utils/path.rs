use std::path::{Component, Path};
use sugar_path::{self, PathSugar};
pub fn normalize_path(path: &str, root: &str) -> String {
  let res = Path::new(&root)
    .relative(Path::new(&path))
    .to_string_lossy()
    .to_string();

  res
}

pub fn uri_to_chunk_name(root: &str, uri: &str) -> String {
  println!("root {:?}", root);
  let path = Path::new(uri);
  let mut relatived = Path::new(root).relative(path);
  let ext = relatived
    .extension()
    .and_then(|ext| ext.to_str())
    .unwrap_or("")
    .to_string();
  relatived.set_extension("");
  let mut name = relatived
    .components()
    .filter(|com| matches!(com, Component::Normal(_)))
    .filter_map(|seg| seg.as_os_str().to_str())
    .intersperse("_")
    .fold(String::new(), |mut acc, seg| {
      acc.push_str(seg);
      acc
    });
  name.push('_');
  name.push_str(&ext);
  name
}
