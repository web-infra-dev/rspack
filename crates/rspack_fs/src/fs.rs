use std::fmt::Debug;

use crate::{ReadableFileSystem, WritableFileSystem};

pub trait FileSystem: ReadableFileSystem + WritableFileSystem + Debug + Sync + Send {}
