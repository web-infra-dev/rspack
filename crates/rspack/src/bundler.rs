use std::collections::HashMap;
use std::collections::HashSet;
use std::path::Path;
use std::path::PathBuf;
use std::sync::Arc;

use anyhow::Result;
use nodejs_resolver::Resolver;
use rayon::prelude::*;
use rspack_core::inject_options;
use rspack_core::Bundle;
use rspack_core::NormalizedBundleOptions;
use rspack_core::{OutputChunk, OutputChunkSourceMap};
use rspack_swc::swc_common;
use sugar_path::PathSugar;
use swc_common::Mark;
use tracing::instrument;

use crate::chunk_spliter::generate_chunks;
use crate::stats::Stats;
use crate::utils::inject_built_in_plugins;
use crate::utils::log::enable_tracing_by_env;
use crate::utils::rayon::init_rayon_thread_poll;
use md4::{Digest, Md4};
pub use rspack_core::finalize::finalize;
use rspack_core::get_swc_compiler;
use rspack_core::Plugin;
use rspack_core::PluginDriver;

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum InternalModuleFormat {
  ES,
  CJS,
  AMD,
  UMD,
}

impl Default for InternalModuleFormat {
  fn default() -> Self {
    InternalModuleFormat::ES
  }
}

pub use rspack_core::Asset;
pub use rspack_core::BundleContext;
pub use rspack_core::BundleMode;
pub use rspack_core::BundleOptions;

#[derive(Debug)]
#[allow(clippy::manual_non_exhaustive)]
pub struct Bundler {
  pub options: Arc<NormalizedBundleOptions>,
  pub plugin_driver: Arc<PluginDriver>,
  pub bundle: Bundle,
  pub resolver: Arc<Resolver>,
  _noop: (),
}

impl Bundler {
  pub fn new(options: BundleOptions, plugins: Vec<Box<dyn Plugin>>) -> Self {
    enable_tracing_by_env();
    init_rayon_thread_poll();
    println!(
      "create bundler with options:\n {:#?} \nplugins:\n {:#?}\n",
      options, plugins
    );
    let normalized_options = Arc::new(inject_options(options));
    let injected_plugins = inject_built_in_plugins(plugins, &normalized_options);
    let (top_level_mark, unresolved_mark) =
      get_swc_compiler().run(|| (Mark::fresh(Mark::root()), Mark::fresh(Mark::root())));

    let ctx: Arc<BundleContext> = Arc::new(BundleContext::new(
      get_swc_compiler(),
      normalized_options.clone(),
      top_level_mark,
      unresolved_mark,
    ));

    let resolver = Arc::new(Resolver::new(normalized_options.resolve.clone()));

    let plugin_driver = Arc::new(PluginDriver::new(
      injected_plugins,
      ctx.clone(),
      resolver.clone(),
    ));

    Self {
      options: normalized_options.clone(),
      plugin_driver: plugin_driver.clone(),
      resolver: resolver.clone(),
      bundle: Bundle::new(normalized_options.clone(), plugin_driver, ctx, resolver),
      _noop: (),
    }
  }

  #[instrument(skip(self))]
  pub async fn build(&mut self, changed_files: Option<Vec<String>>) -> Result<Stats> {
    let start_time = std::time::Instant::now();
    self.plugin_driver.build_start().await?;
    tracing::trace!("start build");

    self.bundle.build_graph(changed_files).await?;

    let chunks = generate_chunks(&mut self.bundle)?;
    self.bundle.chunk_graph = chunks;

    let output = self.render_chunks();

    let mut map = HashMap::default();

    output.into_iter().for_each(|(_, chunk)| {
      map.insert(chunk.file_name.clone(), chunk.entry.clone());
      let mut code = chunk.code;

      match chunk.map {
        OutputChunkSourceMap::Inline(encoded_map) => {
          let comment = "\n//# sourceMappingURL=".to_owned() + &encoded_map;
          code += &comment;
        }
        OutputChunkSourceMap::External(map) => {
          let map_filename = chunk.file_name.clone() + ".map";
          self.bundle.context.assets.lock().unwrap().push(Asset {
            source: map,
            filename: map_filename,
          });
        }
        OutputChunkSourceMap::Linked(map) => {
          let map_filename = chunk.file_name.clone() + ".map";
          self.bundle.context.assets.lock().unwrap().push(Asset {
            source: map,
            filename: map_filename.clone(),
          });
          let comment = "\n//# sourceMappingURL=".to_owned() + &map_filename;
          code += &comment;
        }
        OutputChunkSourceMap::None => (),
      };

      self.bundle.context.assets.lock().unwrap().push(Asset {
        source: code,
        filename: chunk.file_name,
      });
    });

    self.plugin_driver.build_end().await?;
    let end_time = std::time::Instant::now();

    Ok(Stats {
      map,
      start_time,
      end_time,
    })
  }

  pub fn resolve(
    &mut self,
    id: String,
    dir: String,
  ) -> Result<nodejs_resolver::ResolveResult, std::string::String> {
    let base = Path::new(&dir);
    self.resolver.resolve(base, &id)
  }

  #[instrument(skip(self))]
  pub async fn rebuild(
    &mut self,
    changed_file: Vec<String>,
  ) -> Result<Vec<HashMap<String, String>>> {
    tracing::debug!("rebuild because of {:?}", changed_file);
    let changed_files = changed_file;
    let old_modules_uri = self
      .bundle
      .module_graph_container
      .module_graph
      .uris()
      .cloned()
      .filter(|id| !changed_files.contains(id))
      .collect::<HashSet<_>>();

    tracing::trace!("old_modules_id {:?}", old_modules_uri);

    self.bundle.context.assets.lock().unwrap().clear();
    changed_files.iter().for_each(|rd| {
      self
        .bundle
        .module_graph_container
        .module_graph
        .remove_by_uri(rd);
      self.bundle.visited_module_id.remove(rd);
    });

    let Stats { map, .. } = self.build(Some(changed_files)).await?;

    let new_modules_id = self
      .bundle
      .module_graph_container
      .module_graph
      .uris()
      .cloned()
      .collect::<HashSet<_>>();
    let diff_rendered = new_modules_id
      .into_iter()
      .filter(|module_id| !old_modules_uri.contains(module_id))
      .map(|module_id| {
        tracing::trace!("render new added module {:?}", module_id);
        (
          module_id.to_string(),
          self
            .bundle
            .module_graph_container
            .module_graph
            .module_by_uri(&module_id)
            .unwrap()
            .cached_output
            .lock()
            .unwrap()
            .as_ref()
            .unwrap()
            .code
            .clone(),
        )
      })
      .collect();
    Ok(vec![diff_rendered, map])
  }

  pub fn write_assets_to_disk(&self) {
    self
      .bundle
      .context
      .assets
      .lock()
      .unwrap()
      .iter()
      .for_each(|asset| {
        let mut path = PathBuf::from(self.options.outdir.clone());
        path.push(&asset.filename);
        std::fs::create_dir_all(path.resolve().parent().unwrap()).unwrap();
        std::fs::write(path.resolve(), &asset.source).unwrap();
      });
  }

  pub fn render_chunks(&self) -> HashMap<String, OutputChunk> {
    self
      .bundle
      .chunk_graph
      .id_to_chunk()
      .par_iter()
      .map(|(_chunk_id, chunk)| {
        let mut chunk = chunk.render(&self.bundle);
        if chunk.file_name.contains("[contenthash]") {
          let mut hasher = Md4::new();
          hasher.update(&chunk.code);
          let content_hash = hasher.finalize();
          chunk.file_name = chunk
            .file_name
            .replace("[contenthash]", &format!("{:x}", content_hash));
        }

        (chunk.file_name.clone(), chunk)
      })
      .collect()
  }
}
