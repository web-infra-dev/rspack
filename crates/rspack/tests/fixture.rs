mod testing {
  use async_trait::async_trait;
  use rspack::bundler::{BundleContext, BundleOptions, Bundler};
  use rspack_core::{Loader, ResolveOption};
  use rspack_core::{
    Plugin, PluginLoadHookOutput, PluginResolveHookOutput, PluginTransformHookOutput,
  };

  use rspack_swc::swc_ecma_ast;
  use serde_json::Value;
  use std::collections::HashMap;
  use std::env;
  use std::fs;
  use std::path::Path;
  use std::sync::atomic::AtomicBool;
  use std::sync::Arc;
  use std::sync::Once;

  static INIT: Once = Once::new();

  fn compile(fixture_path: &str, plugins: Vec<Box<dyn Plugin>>) -> Bundler {
    INIT.call_once(|| {
      let default_panic = std::panic::take_hook();
      std::panic::set_hook(Box::new(move |info| {
        default_panic(info);
        std::process::exit(1);
      }));
    });
    compile_with_options(fixture_path, Default::default(), plugins)
  }

  fn compile_with_options(
    fixture_path: &str,
    options: BundleOptions,
    plugins: Vec<Box<dyn Plugin>>,
  ) -> Bundler {
    compile_with_options_inner(fixture_path, options, plugins)
  }

  #[tokio::main]
  async fn compile_with_options_inner(
    fixture_path: &str,
    options: BundleOptions,
    plugins: Vec<Box<dyn Plugin>>,
  ) -> Bundler {
    let dir = Path::new(env!("CARGO_MANIFEST_DIR"));
    let fixtures_dir = dir.join("fixtures").join(fixture_path);
    let pkg_path = fixtures_dir.join("package.json");
    let pkg_content = fs::read_to_string(pkg_path);
    let mut pkg: Value = Value::default();
    if pkg_content.is_ok() {
      pkg = serde_json::from_str(&pkg_content.unwrap()).unwrap();
    }
    // use pkg.rspack.entry if set otherwise use index.js as entry
    let pkg_entry = pkg["rspack"].clone()["entry"].clone();
    let entry = {
      if pkg_entry.is_object() {
        let obj: HashMap<String, String> = serde_json::from_value(pkg_entry).unwrap();
        obj
          .into_iter()
          .map(|(_id, value)| {
            let resolve_path = fixtures_dir.join(value).display().to_string();
            return resolve_path;
          })
          .collect()
      } else {
        let default_entry = fixtures_dir.join("index.js").to_str().unwrap().to_string();
        vec![default_entry]
      }
    };
    let dist = fixtures_dir.join("dist");
    println!("entry: {:?}", entry);
    let mut bundler = Bundler::new(
      BundleOptions {
        entries: entry,
        outdir: dist.to_str().unwrap().to_string(),
        ..options
      },
      plugins,
    );
    bundler.build().await;
    bundler.write_assets_to_disk();
    bundler
  }

  #[test]
  fn single_entry() {
    compile("single-entry", vec![]);
  }

  #[test]
  fn multi_entry() {
    compile("multi-entry", vec![]);
  }

  #[test]
  fn cycle_dep() {
    compile("cycle-dep", vec![]);
  }

  #[derive(Debug)]
  struct TestPlugin {
    call_resolve: Arc<AtomicBool>,
    call_load: Arc<AtomicBool>,
    call_transform: Arc<AtomicBool>,
  }

  #[async_trait]
  impl Plugin for TestPlugin {
    fn name(&self) -> &'static str {
      "rspack_test"
    }

    async fn resolve(
      &self,
      _ctx: &BundleContext,
      _id: &str,
      _importer: Option<&str>,
    ) -> PluginResolveHookOutput {
      self
        .call_resolve
        .store(true, std::sync::atomic::Ordering::SeqCst);
      None
    }

    #[inline]
    async fn load(&self, _ctx: &BundleContext, _id: &str) -> PluginLoadHookOutput {
      self
        .call_load
        .store(true, std::sync::atomic::Ordering::SeqCst);
      None
    }

    #[inline]
    fn transform(
      &self,
      _ctx: &BundleContext,
      _path: &Path,
      ast: swc_ecma_ast::Module,
    ) -> PluginTransformHookOutput {
      self
        .call_transform
        .store(true, std::sync::atomic::Ordering::SeqCst);
      ast
    }
  }

  #[test]
  fn plugin_test() {
    let call_resolve: Arc<AtomicBool> = Default::default();
    let call_load: Arc<AtomicBool> = Default::default();
    let call_transform: Arc<AtomicBool> = Default::default();
    let test_plugin = Box::new(TestPlugin {
      call_resolve: call_resolve.clone(),
      call_load: call_load.clone(),
      call_transform: call_transform.clone(),
    });
    compile("single-entry", vec![test_plugin]);
    assert!(call_load.load(std::sync::atomic::Ordering::SeqCst));
    assert!(call_resolve.load(std::sync::atomic::Ordering::SeqCst));
    assert!(call_transform.load(std::sync::atomic::Ordering::SeqCst));
  }

  #[test]
  fn dynamic_import() {
    compile("dynamic-import", vec![]);
  }

  #[test]
  #[ignore]
  fn basic_css() {
    let bundler = compile("basic-css", vec![]);
    assert!(bundler
      .plugin_driver
      .plugins
      .iter()
      .find(|plugin| plugin.name() == rspack_plugin_css::plugin::PLUGIN_NAME)
      .is_some())
  }

  #[test]
  #[ignore = "not support npm yet"]
  fn npm() {
    compile("npm", vec![]);
  }

  #[test]
  fn cjs() {
    compile("cjs", vec![]);
  }

  #[test]
  fn css_bundle_test() {
    compile("css", vec![]);
  }

  #[test]
  fn disable_code_splitting() {
    let bundler = compile_with_options(
      "basic",
      BundleOptions {
        code_splitting: false,
        ..Default::default()
      },
      vec![],
    );
    let chunk_len = bundler.ctx.assets.lock().unwrap().len();
    assert_eq!(chunk_len, 2);
  }

  #[test]
  fn enable_code_splitting() {
    let bundler = compile_with_options(
      "basic",
      BundleOptions {
        code_splitting: true,
        ..Default::default()
      },
      vec![],
    );
    let chunk_len = bundler.ctx.assets.lock().unwrap().len();
    assert_eq!(chunk_len, 4);
  }

  #[test]
  fn basic_ts() {
    compile("basic-ts", vec![]);
  }
  #[test]
  fn splitting() {
    compile("code-splitting", vec![]);
  }

  #[test]
  fn loader() {
    compile_with_options(
      "loader",
      BundleOptions {
        loader: Some(
          vec![
            ("svg".to_string(), Loader::DataURI),
            // Json is supported by default
            // ("json".to_string(), Loader::Json),
            ("txt".to_string(), Loader::Text),
          ]
          .into_iter()
          .collect(),
        ),
        ..Default::default()
      },
      vec![],
    );
  }

  #[test]
  fn alias() {
    let bundler = compile_with_options(
      "alias",
      BundleOptions {
        resolve: ResolveOption {
          alias: vec![
            ("./wrong".to_string(), Some("./ok".to_string())),
            ("@/".to_string(), Some("./src/".to_string())),
          ],
          ..Default::default()
        },
        ..Default::default()
      },
      vec![],
    );
    let assets = bundler.ctx.assets.lock().unwrap();
    let dist = assets.get(0).unwrap();
    let source = &dist.source;
    println!("assets {:#?}", assets);
    assert!(!source.contains("wrong.js"));
    assert!(!source.contains("@"));
    assert!(source.contains("ok.js"));
    assert!(source.contains("at.js"));
  }

  #[test]

  fn stack_overflow_mockjs() {
    compile("stack_overflow_mockjs", vec![]);
  }
}
