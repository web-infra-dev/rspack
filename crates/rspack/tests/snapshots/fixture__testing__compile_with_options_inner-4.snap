---
source: crates/rspack/tests/fixture.rs
expression: bundler.assets()
---
[
    Asset {
        source: "rs.define(\"/Users/yangjian/github/rspack/crates/rspack/fixtures/basic/b.js\", function(require, module, exports) {\n    \"use strict\";\n    Object.defineProperty(exports, \"__esModule\", {\n        value: true\n    });\n    exports.b = void 0;\n    var b = 1;\n    exports.b = b;\n});\nrs.define(\"/Users/yangjian/github/rspack/crates/rspack/fixtures/basic/entry-b.js\", function(require, module, exports) {\n    \"use strict\";\n    var _b = require(\"/Users/yangjian/github/rspack/crates/rspack/fixtures/basic/b.js\");\n    var _shared = require(\"/Users/yangjian/github/rspack/crates/rspack/fixtures/basic/shared.js\");\n    console.log(_b.b, _shared.shared);\n});\nrs.require(\"/Users/yangjian/github/rspack/crates/rspack/fixtures/basic/entry-b.js\");\nrs.define(\"/Users/yangjian/github/rspack/crates/rspack/fixtures/basic/shared.js\", function(require, module, exports) {\n    \"use strict\";\n    Object.defineProperty(exports, \"__esModule\", {\n        value: true\n    });\n    exports.shared = void 0;\n    var shared = 1;\n    exports.shared = shared;\n});\n",
        filename: "entry-b.js",
    },
    Asset {
        source: "rs.define(\"/Users/yangjian/github/rspack/crates/rspack/fixtures/basic/a.js\", function(require, module, exports) {\n    \"use strict\";\n    Object.defineProperty(exports, \"__esModule\", {\n        value: true\n    });\n    exports.a = void 0;\n    var a = 1;\n    exports.a = a;\n});\nrs.define(\"/Users/yangjian/github/rspack/crates/rspack/fixtures/basic/shared.js\", function(require, module, exports) {\n    \"use strict\";\n    Object.defineProperty(exports, \"__esModule\", {\n        value: true\n    });\n    exports.shared = void 0;\n    var shared = 1;\n    exports.shared = shared;\n});\nrs.define(\"/Users/yangjian/github/rspack/crates/rspack/fixtures/basic/entry-a.js\", function(require, module, exports) {\n    \"use strict\";\n    function _interopRequireWildcard(obj) {\n        if (obj && obj.__esModule) {\n            return obj;\n        } else {\n            var newObj = {};\n            if (obj != null) {\n                for(var key in obj){\n                    if (Object.prototype.hasOwnProperty.call(obj, key)) {\n                        var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {};\n                        if (desc.get || desc.set) {\n                            Object.defineProperty(newObj, key, desc);\n                        } else {\n                            newObj[key] = obj[key];\n                        }\n                    }\n                }\n            }\n            newObj.default = obj;\n            return newObj;\n        }\n    }\n    var _a = require(\"/Users/yangjian/github/rspack/crates/rspack/fixtures/basic/a.js\");\n    var _shared = require(\"/Users/yangjian/github/rspack/crates/rspack/fixtures/basic/shared.js\");\n    console.log(_a.a, _shared.shared);\n    Promise.resolve().then(function() {\n        return _interopRequireWildcard(require(\"./asynced\"));\n    }).then(console.log);\n});\nrs.require(\"/Users/yangjian/github/rspack/crates/rspack/fixtures/basic/entry-a.js\");\nrs.define(\"/Users/yangjian/github/rspack/crates/rspack/fixtures/basic/asynced.js\", function(require, module, exports) {\n    \"use strict\";\n    Object.defineProperty(exports, \"__esModule\", {\n        value: true\n    });\n    exports.asynced = void 0;\n    var asynced = 1;\n    exports.asynced = asynced;\n});\n",
        filename: "entry-a.js",
    },
]
